File: CourseOverlap.jl
module CourseOverlap

include("Parse.jl")
include("Utils.jl")

import .Parse: get_plans
import .Utils: writerow

plans = get_plans()

open("./files/course_overlap.csv", "w") do file
  writerow(file, [
    "Year",
    "Base major",
    "Other major",
    "Percent of base major's courses in other major",
  ])

  for year in 2015:2050
    if year ∉ keys(plans)
      break
    end
    curricula = Dict(
      major => Set([
        course
        for term in degree_plans[first(
          college
          for college in ["TH", "WA", "SN", "MU", "FI", "RE", "SI", "EI"]
          if college ∈ keys(degree_plans) && !(college == "SN" && year < 2020)
        )]
        for course in term
        if course.for_major
      ])
      for (major, degree_plans) in plans[year]
    )
    majors = sort(collect(keys(curricula)))
    for base in majors
      for other in majors
        writerow(file, [
          string(year),
          base,
          other,
          string(length(curricula[base] ∩ curricula[other]) / length(curricula[base]))
        ])
      end
    end
  end
end

end

File: course_overlap.py
"""
python3 course_overlap.py
"""

from parse import major_plans
from util import CsvWriter

HEADER = [
    "Year",
    "Base major",
    "Other major",
    "Percent of base major's courses in other major",
]


def print_year(writer: CsvWriter, year: int) -> None:
    majors = major_plans(year)
    if majors == {}:
        return

    curricula = {
        major: {course for course in plans.curriculum() if course.for_major}
        for major, plans in majors.items()
    }
    major_codes = sorted(majors.keys())
    for base in major_codes:
        for other in major_codes:
            writer.row(
                str(year),
                base,
                other,
                str(
                    len(curricula[base] & curricula[other]) / len(curricula[base])
                    if curricula[base]
                    else 0
                ),
            )


def main() -> None:
    with open("./files/course_overlap_py.csv", "w") as file:
        writer = CsvWriter(len(HEADER), file)
        writer.row(*HEADER)

        for year in range(2015, 2050):
            print_year(writer, year)


if __name__ == "__main__":
    main()
